src/plugins/data/public/query/query_string/query_string_manager.ts

```diff
@@ -57,6 +57,28 @@ export class QueryStringManager {
     this.queryHistory = createHistory({ storage: this.sessionStorage });
     this.datasetService = new DatasetService(uiSettings, this.sessionStorage);
     this.languageService = new LanguageService(this.defaultSearchInterceptor, this.storage);
+
+    this.uiSettings.get$(UI_SETTINGS.QUERY_ENHANCEMENTS_ENABLED).subscribe((isEnabled) => {
```

Comment:
This isn't safe enough. It should be something on page load ideally that checks for the flag and does any cleanup actions. Since we require a page reload for the flag, that should take care of the change listener missing the change and letting the bad state persist

Reply:
Is it possible that the setting changes before subscription set up or already in a bad state (ex. SQL) when the subscription is established?